Enable local clang-tidy lint

1ntEgr8 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61121**

This change enables the make target to run clang-tidy locally

Test Plan:

Run this command
```
make clang-tidy
```
This should run `clang-tidy` on the paths and filters specified in `tools/linter/clang_tidy/__main__.py`

Quicklint
```
make quicklint
```
This should report "No files detected" if no c/cpp files are altered.

Differential Revision: [D29598927](https://our.internmc.facebook.com/intern/diff/D29598927)

1ntEgr8 committed:
Enable local clang-tidy lint

This change enables the make target to run clang-tidy locally

[ghstack-poisoned]

1ntEgr8 committed:
Update on "Enable local clang-tidy lint"


This change enables the make target to run clang-tidy locally

Test Plan:

Run this command
```
make clang_tidy
```
This should run `clang-tidy` on the paths and filters specified in `tools/linter/clang_tidy/__main__.py`

Quicklint
```
make quicklint
```
This should report "No files detected" if no c/cpp files are altered.

[ghstack-poisoned]

1ntEgr8 committed:
Update on "Enable local clang-tidy lint"


This change enables the make target to run clang-tidy locally

Test Plan:

Run this command
```
make clang_tidy
```
This should run `clang-tidy` on the paths and filters specified in `tools/linter/clang_tidy/__main__.py`

Quicklint
```
make quicklint
```
This should report "No files detected" if no c/cpp files are altered.

[ghstack-poisoned]

1ntEgr8 committed:
Update on "Enable local clang-tidy lint"


This change enables the make target to run clang-tidy locally

Test Plan:

Run this command
```
make clang_tidy
```
This should run `clang-tidy` on the paths and filters specified in `tools/linter/clang_tidy/__main__.py`

Quicklint
```
make quicklint
```
This should report "No files detected" if no c/cpp files are altered.

[ghstack-poisoned]

janeyx99 commented on Makefile:
```diff
@@ -105,8 +105,10 @@ cmakelint:
 		--step 'Run cmakelint'
 
 clang_tidy:
-	echo "clang-tidy local lint is not yet implemented"
```
🎉

1ntEgr8 committed:
Update on "Enable local clang-tidy lint"


This change enables the make target to run clang-tidy locally

Test Plan:

Run this command
```
make clang_tidy
```
This should run `clang-tidy` on the paths and filters specified in `tools/linter/clang_tidy/__main__.py`

Quicklint
```
make quicklint
```
This should report "No files detected" if no c/cpp files are altered.

[ghstack-poisoned]

janeyx99 commented on Makefile:
```diff
@@ -105,8 +105,10 @@ cmakelint:
 		--step 'Run cmakelint'
 
 clang_tidy:
```
could this be clang-tidy to be consistent with the previous GHA workflow name? 
```suggestion
clang-tidy:
```

janeyx99 reviewed (approved):
nice :D 

don't land before fixing flake8 tho

driazati commented on tools/linter/clang_tidy/__main__.py:
```diff
@@ -68,7 +71,13 @@ def parse_args() -> argparse.Namespace:
+        "--diff-file",
+        help="File containing diff to use for determining files to lint and line filters",
+    )
+    parser.add_argument(
```
I'd prefer if we kept `actions_local_runner.py` as a main dispatcher for local lints so we can avoid duplicating stuff like `--changed-only`, it could dispatch out to this script but could you add it as an explicit job here?

https://github.com/pytorch/pytorch/blob/master/tools/actions_local_runner.py#L407

If you do that and implment `quick()`, `full()`, and `filter_files()` this should all come for free. It should work to pass the files list through manually with the `--path` option here so we don't have to add any new ones.

driazati commented on Makefile:
```diff
@@ -105,8 +105,10 @@ cmakelint:
 clang_tidy:
-	echo "clang-tidy local lint is not yet implemented"
-	exit 1
+	@$(PYTHON) tools/linter/clang_tidy \
```
This can call the local runner instead which can encode all the args like `--parallel` and the `--clang-tidy-exe`

1ntEgr8 committed:
Update on "Enable local clang-tidy lint"


This change enables the make target to run clang-tidy locally

Test Plan:

Run this command
```
make clang_tidy
```
This should run `clang-tidy` on the paths and filters specified in `tools/linter/clang_tidy/__main__.py`

Quicklint
```
make quicklint
```
This should report "No files detected" if no c/cpp files are altered.

[ghstack-poisoned]

1ntEgr8 committed:
Update on "Enable local clang-tidy lint"


This change enables the make target to run clang-tidy locally

Test Plan:

Run this command
```
make clang_tidy
```
This should run `clang-tidy` on the paths and filters specified in `tools/linter/clang_tidy/__main__.py`

Quicklint
```
make quicklint
```
This should report "No files detected" if no c/cpp files are altered.

[ghstack-poisoned]

1ntEgr8 committed:
Update on "Enable local clang-tidy lint"


This change enables the make target to run clang-tidy locally

Test Plan:

Run this command
```
make clang_tidy
```
This should run `clang-tidy` on the paths and filters specified in `tools/linter/clang_tidy/__main__.py`

Quicklint
```
make quicklint
```
This should report "No files detected" if no c/cpp files are altered.

[ghstack-poisoned]

1ntEgr8 committed:
Update on "Enable local clang-tidy lint"


This change enables the make target to run clang-tidy locally

Test Plan:

Run this command
```
make clang_tidy
```
This should run `clang-tidy` on the paths and filters specified in `tools/linter/clang_tidy/__main__.py`

Quicklint
```
make quicklint
```
This should report "No files detected" if no c/cpp files are altered.

[ghstack-poisoned]

1ntEgr8 committed:
Update on "Enable local clang-tidy lint"


This change enables the make target to run clang-tidy locally

Test Plan:

Run this command
```
make clang_tidy
```
This should run `clang-tidy` on the paths and filters specified in `tools/linter/clang_tidy/__main__.py`

Quicklint
```
make quicklint
```
This should report "No files detected" if no c/cpp files are altered.

[ghstack-poisoned]

1ntEgr8 committed:
Update on "Enable local clang-tidy lint"


This change enables the make target to run clang-tidy locally

Test Plan:

Run this command
```
make clang_tidy
```
This should run `clang-tidy` on the paths and filters specified in `tools/linter/clang_tidy/__main__.py`

Quicklint
```
make quicklint
```
This should report "No files detected" if no c/cpp files are altered.

[ghstack-poisoned]

driazati commented on tools/actions_local_runner.py:
```diff
@@ -289,6 +290,35 @@ async def full(self) -> None:
+        await shell_cmd(
+            [
+                sys.executable,
+                "tools/linter/clang_tidy",
```
It looks like this script always returns `0` as its exit code, so even if there are failures the runner still reports that it passed (and it doesn't print the output when the command passes). Can you fix up the clang-tidy script to thread through the return code + unify them from parallel runs?

1ntEgr8 replied:
`clang-tidy` returns a zero exit code even if there are warnings. To resolve this comment, I'm using the `extract_warnings` method to detect if there are any warnings, and then return the number of warnings as the exit code.

driazati commented on tools/linter/clang_tidy/__main__.py:
```diff
@@ -124,6 +163,20 @@ def main() -> None:
+        msg = (
+            "Could not find 'clang-tidy' binary\n"
+            "You can install it by running:\n"
+            "   python3 tools/linter/install/clang_tidy.py"
```
Can you throw this into the `make setup_lint` command too?

driazati commented on tools/linter/clang_tidy/run.py:
```diff
@@ -256,16 +270,54 @@ def apply_nolint(fname: str, warnings: Dict[int, Set[str]]) -> None:
         f.write("".join(lines))
 
 
+def git(args: List[str]) -> str:
```
nit: unused function?

driazati reviewed (approved):
lgtm, it should be fine once we get the clang-tidy runner to properly report its return code

1ntEgr8 committed:
Update on "Enable local clang-tidy lint"


This change enables the make target to run clang-tidy locally

Test Plan:

Run this command
```
make clang-tidy
```
This should run `clang-tidy` on the paths and filters specified in `tools/linter/clang_tidy/__main__.py`

Quicklint
```
make quicklint
```
This should report "No files detected" if no c/cpp files are altered.

[ghstack-poisoned]

1ntEgr8 committed:
Update on "Enable local clang-tidy lint"


This change enables the make target to run clang-tidy locally

Test Plan:

Run this command
```
make clang-tidy
```
This should run `clang-tidy` on the paths and filters specified in `tools/linter/clang_tidy/__main__.py`

Quicklint
```
make quicklint
```
This should report "No files detected" if no c/cpp files are altered.

[ghstack-poisoned]

1ntEgr8 committed:
Update on "Enable local clang-tidy lint"


This change enables the make target to run clang-tidy locally

Test Plan:

Run this command
```
make clang-tidy
```
This should run `clang-tidy` on the paths and filters specified in `tools/linter/clang_tidy/__main__.py`

Quicklint
```
make quicklint
```
This should report "No files detected" if no c/cpp files are altered.

[ghstack-poisoned]

1ntEgr8 committed:
Update on "Enable local clang-tidy lint"


This change enables the make target to run clang-tidy locally

Test Plan:

Run this command
```
make clang-tidy
```
This should run `clang-tidy` on the paths and filters specified in `tools/linter/clang_tidy/__main__.py`

Quicklint
```
make quicklint
```
This should report "No files detected" if no c/cpp files are altered.

[ghstack-poisoned]

1ntEgr8 commented:
@1ntEgr8 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29598927).

1ntEgr8 committed:
Update on "Enable local clang-tidy lint"


This change enables the make target to run clang-tidy locally

Test Plan:

Run this command
```
make clang-tidy
```
This should run `clang-tidy` on the paths and filters specified in `tools/linter/clang_tidy/__main__.py`

Quicklint
```
make quicklint
```
This should report "No files detected" if no c/cpp files are altered.

Differential Revision: [D29598927](https://our.internmc.facebook.com/intern/diff/D29598927)

[ghstack-poisoned]

1ntEgr8 commented:
@1ntEgr8 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29598927).
